src/transformers/models/grounding_dino/modeling_grounding_dino.py

```diff
@@ -0,0 +1,3714 @@
+            text_encoder_hidden_states=encoder_outputs[1],
+            text_encoder_attention_mask=~text_token_mask,
+            reference_points=reference_points,
+            spatial_shapes=spatial_shapes,
+            level_start_index=level_start_index,
+            valid_ratios=valid_ratios,
+            self_attn_mask=None,
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+            return_dict=return_dict,
+        )
+
+        if not return_dict:
+            enc_outputs = tuple(value for value in [enc_outputs_class, enc_outputs_coord_logits] if value is not None)
+            tuple_outputs = (init_reference_points,) + decoder_outputs + encoder_outputs + enc_outputs
+
+            return tuple_outputs
+
+        return GroundingDinoModelOutput(
+            init_reference_points=init_reference_points,
+            last_hidden_state=decoder_outputs.last_hidden_state,
+            intermediate_hidden_states=decoder_outputs.intermediate_hidden_states,
+            intermediate_reference_points=decoder_outputs.intermediate_reference_points,
+            decoder_hidden_states=decoder_outputs.hidden_states,
+            decoder_attentions=decoder_outputs.attentions,
+            encoder_last_hidden_state_vision=encoder_outputs.last_hidden_state_vision,
+            encoder_last_hidden_state_text=encoder_outputs.last_hidden_state_text,
+            encoder_hidden_states_vision=encoder_outputs.hidden_states_vision,
+            encoder_hidden_states_text=encoder_outputs.hidden_states_text,
+            encoder_attentions=encoder_outputs.attentions,
+            enc_outputs_class=enc_outputs_class,
+            enc_outputs_coord_logits=enc_outputs_coord_logits,
+        )
+
+
+@add_start_docstrings(
+    """
+    Grounding DINO Model (consisting of a backbone and encoder-decoder Transformer) with object detection heads on top,
+    for tasks such as COCO detection.
+    """,
+    GROUNDING_DINO_START_DOCSTRING,
+)
+class GroundingDinoForObjectDetection(GroundingDinoPreTrainedModel):
+    # When using clones, all layers > 0 will be clones, but layer 0 *is* required
+    _tied_weights_keys = [r"bbox_embed\.[1-9]\d*"]
+
+    def __init__(self, config: GroundingDinoConfig):
+        super().__init__(config)
+
+        # Deformable DETR encoder-decoder model
+        self.model = GroundingDinoModel(config)
+
+        # Detection heads on top
+        _class_embed = GroundingDinoContrastiveEmbedding(config)
+        _bbox_embed = GroundingDinoMLPPredictionHead(
+            input_dim=config.d_model, hidden_dim=config.d_model, output_dim=4, num_layers=3
+        )
+
+        if config.decoder_bbox_embed_share:
+            self.bbox_embed = nn.ModuleList([_bbox_embed for _ in range(config.decoder_layers)])
+        else:
+            self.bbox_embed = _get_clones(_bbox_embed, config.decoder_layers)
+        self.class_embed = nn.ModuleList([_class_embed for _ in range(config.decoder_layers)])
```

Comment:
Makes sense, not sure if it is worth looking at `DeformableDetr` then since this came from `add model-like deformable-detr`

Reply:
Ha! Well, you can tell I didn't review for that model ;)

Nevertheless, two wrongs don't make a right, let's do it better for this one.

Reply:
Now that I was about to fix this I have a question. What do you mean by "memory burden" in the case we share the layer? Because at first I thought you were saying to instead of doing:

```python
self.bbox_embed = nn.ModueList([_bbox_embed for I in range(...)])
...
def forward(...):
...
for bbox in self.bbox_embed:
  x = bbox(x)
```

to do this:

```python
self.bbox_embed = ...

def forward:
...
for _ in range(config.decoder_layers):
  x = self.bbox_embed(x)
```

But if that is the case I would imagine that it doesn't matter, right? Memory usage should be the same since there's only one reference to parameters. Consider this example:

```python
import sys

from torch import nn

num_layers = 5

layer = nn.Linear(100, 100)
layers = nn.ModuleList([layer for _  in range(num_layers)])
layers_2 = nn.ModuleList([nn.Linear(100, 100) for _ in range(num_layers)])

print(sys.getsizeof(list(layer.parameters()))) # 120
print(sys.getsizeof(list(layers.parameters()))) # 120
print(sys.getsizeof(list(layers_2.parameters()))) # 184

def print_named_parameters(m):
    for name, param in m.named_parameters():
        print(name, param.shape)

print_named_parameters(layer)
# weight torch.Size([100, 100])
# bias torch.Size([100])
print_named_parameters(layers)
# 0.weight torch.Size([100, 100])
# 0.bias torch.Size([100])
print_named_parameters(layers_2)
# 0.weight torch.Size([100, 100])
# 0.bias torch.Size([100])
# 1.weight torch.Size([100, 100])
# 1.bias torch.Size([100])
# 2.weight torch.Size([100, 100])
# 2.bias torch.Size([100])
# 3.weight torch.Size([100, 100])
# 3.bias torch.Size([100])
# 4.weight torch.Size([100, 100])
# 4.bias torch.Size([100])
```

Let me know if I overlooked anything else

Reply:
What I was meaning is that, if the embedding is really shared, then we don't need a list at all i.e., instead of:

```self.bbox_embed = nn.ModueList([_bbox_embed for I in range(...)])```

we can just have

```self.bbox_embed = _bbox_embed```

which we call n times. 

With regards to my previous comment: 
>If we're not sharing then we're using a copy hack to paper over not instantiating the class properly at the start

I was trying to say, if these parameters are really shared, then this list comprehension / copying isn't needed. Or, they are not shared, in which case we shouldn't be using the deep copy logic that was in `_get_clones` at all, but instead properly creating the classes.

Reply:
Right now we aren't using the `_get_clones` anymore, but I would still favor using the `ModuleList` for the case where parameters are shared since this maintains consistency when parameters are not shared and doesn't add complexity to `forward`, moreover since there's no (clear) disadvantage of using `ModuleList` I would leave as it is